File: platforms/ios/Dezel/Dezel/Core/JavaScriptObjectBuilder.swift
/**
 * @class JavaScriptObjectBuilder
 * @since 0.1.0
 * @hidden
 */
internal final class JavaScriptObjectBuilder: JavaScriptBuilder {

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
     * @method build
	 * @since 0.1.0
	 * @hidden
     */
	internal class func build(_ context: JavaScriptContext, template: AnyClass) -> JavaScriptValue {

		guard let klass = template as? JavaScriptObject.Type else {
			fatalError("Error creating object instance. Expected JavaScriptObject, found \(template)")
		}

		let value = context.createEmptyObject()

		self.forEach(template, callback: { (name, type, sel, imp) -> Void in

			if (type == .function) {
				DLValueDefineFunction(context.handle, value.handle, name, JavaScriptClassFunctionWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .getter) {
				DLValueDefinePropertyGetter(context.handle, value.handle, name, JavaScriptClassGetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .setter) {
				DLValueDefinePropertySetter(context.handle, value.handle, name, JavaScriptClassSetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

		})

		let instance = klass.init(context: context)
		instance.reset(value.handle)

		DLValueSetAssociatedObject(context.handle, value.handle, UnsafeMutableRawPointer(value: instance))

		return value
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptObjectBuilder.kt
package ca.logaritm.dezel.core

import android.util.Log
import ca.logaritm.dezel.core.external.JavaScriptObjectBuilderExternal
import ca.logaritm.dezel.core.external.JavaScriptValueExternal
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * @class JavaScriptObjectBuilder
 * @super JavaScriptBuilder
 * @since 0.1.0
 */
public class JavaScriptObjectBuilder : JavaScriptBuilder() {

	companion object {

		/**
		 * @method build
		 * @since 0.1.0
		 */
		public fun build(context: JavaScriptContext, template: Class<*>): JavaScriptValue {

			val value = context.createEmptyObject()

			forEach(template, fun(name: String, type: Type, method: Method) {

				if (type == Type.FUNCTION) {
					createFunction(context, value.handle, name, method.name, template)
					return
				}

				if (type == Type.GETTER) {
					createGetter(context, value.handle, name, method.name, template)
					return
				}

				if (type == Type.SETTER) {
					createSetter(context, value.handle, name, method.name, template)
					return
				}

			})

			try {

				val instance = template.getConstructor(JavaScriptContext::class.java).newInstance(context)
				if (instance is JavaScriptObject) {
					instance.reset(value.handle)
				}

				JavaScriptValueExternal.setAssociatedObject(context.handle, value.handle, instance)

			} catch (e: InstantiationException) {
				Log.e("DEZEL", "JavaScriptObjectBuilder Error", e)
			} catch (e: IllegalAccessException) {
				Log.e("DEZEL", "JavaScriptObjectBuilder Error", e)
			} catch (e: InvocationTargetException) {
				Log.e("DEZEL", "JavaScriptObjectBuilder Error", e)
			} catch (e: NoSuchMethodException) {
				Log.e("DEZEL", "JavaScriptObjectBuilder Error", e)
			}

			return value
		}

		/**
		 * @method createFunction
		 * @since 0.1.0
		 * @hidden
		 */
		private fun createFunction(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptObjectBuilderExternal.createFunction(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createGetter
		 * @since 0.1.0
		 * @hidden
		 */
		private fun createGetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptObjectBuilderExternal.createGetter(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createSetter
		 * @since 0.1.0
		 * @hidden
		 */
		private fun createSetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptObjectBuilderExternal.createSetter(context.handle, target, name, method, cls, context)
		}
	}
}
